File: backend/cpp_module/parser.cpp
// backend/cpp_module/parser.cpp
#include <iostream>
#include <string>
#include <regex>
//#include <sstream>
//#include <ctime>

using namespace std;

// Структура для хранения распаршенного лога
struct ParsedLog {
    string timestamp;
    string level;      // "Error", "Warning", "Info"
    string source;     // "System", "Application"
    string message;
    string raw_line;
    bool is_error;
};

// Функция для парсинга одной строки лога Windows
ParsedLog parseWindowsLogLine(const string& line) {
    ParsedLog result;
    result.raw_line = line;
    result.is_error = false;
    
    // Пример: "Error 2025-04-06T10:15:30.000000000Z Service Control Manager 7036 Service started"
    regex pattern(R"((\w+)\s+(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z)\s+([\w\s]+?)\s+(\d+)\s+(.+))");
    
    smatch matches;
    if (regex_search(line, matches, pattern)) {
        result.level = matches[1];           // Error/Warning/Info
        result.timestamp = matches[2];       // ISO timestamp
        result.source = matches[3];          // Source (e.g., "Service Control Manager")
        result.message = matches[5];         // Message
        
        if (result.level == "Error") {
            result.is_error = true;
        }
    } else {
        // Если не подошел паттерн, пробуем альтернативный
        result.level = "Info";
        result.source = "Unknown";
        result.message = line;
        result.is_error = false;
    }
    
    return result;
}

// Функция для проверки, является ли событие критическим
bool isCritical(const ParsedLog& log) {
    if (log.is_error) return true;
    
    // Ключевые слова для алертов
    vector<string> keywords = {"fail", "crash", "timeout", "out of memory", "disk full"};
    for (const auto& kw : keywords) {
        if (log.message.find(kw) != string::npos) {
            return true;
        }
    }
    return false;
}

// Главная функция: читает из stdin, пишет в stdout
int main() {
    string line;
    
    // Читаем строки из stdin (пока не закроют)
    while (getline(cin, line)) {
        // Парсим строку
        ParsedLog parsed = parseWindowsLogLine(line);
        
        // Формируем JSON-вывод для Go
        cout << "PARSED|" 
             << parsed.timestamp << "|"
             << parsed.level << "|"
             << parsed.source << "|"
             << parsed.message << "|"
             << (parsed.is_error ? "true" : "false") << "|"
             << (isCritical(parsed) ? "true" : "false")
             << endl;
    }
    
    return 0;
}

File: parser_module/parser.py
# parser_module/parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import json
import re
from datetime import datetime

class WindowsLogParser:
    """Парсер Windows Event Logs"""
    
    def __init__(self):
        pass
    
    def parse_event(self, raw_event: str) -> dict:
        """Парсит одно событие Windows Event Log"""
        result = {
            'log_name': '',
            'timestamp': '',
            'level': 'Info',
            'source': '',
            'event_id': '',
            'message': '',
            'is_error': False,
            'is_critical': False,
            'raw': raw_event[:200]  # первые 200 символов для отладки
        }
        
        # Определяем тип лога по содержимому
        if 'Log Name:' in raw_event:
            log_match = re.search(r'Log Name:\s*(\S+)', raw_event)
            if log_match:
                result['log_name'] = log_match.group(1)
        
        # Временная метка
        date_match = re.search(r'Date:\s*([0-9-]+\s+[0-9:]+)', raw_event)
        if date_match:
            result['timestamp'] = date_match.group(1)
        
        # Уровень события (Level)
        level_match = re.search(r'Level:\s*(\w+)', raw_event)
        if level_match:
            level = level_match.group(1)
            if 'Error' in level or 'Ошибк' in level:
                result['level'] = 'Error'
                result['is_error'] = True
            elif 'Warning' in level or 'Предупрежд' in level:
                result['level'] = 'Warning'
        
        # Источник (Source)
        source_match = re.search(r'Source:\s*(\S+)', raw_event)
        if source_match:
            result['source'] = source_match.group(1)
        
        # Event ID
        eid_match = re.search(r'Event ID:\s*(\d+)', raw_event)
        if eid_match:
            result['event_id'] = eid_match.group(1)
        
        # Описание (Description)
        desc_match = re.search(r'Description:\s*(.+?)(?=\n\s*\n|\Z)', raw_event, re.DOTALL)
        if desc_match:
            desc = desc_match.group(1).strip()
            result['message'] = desc[:500]  # ограничиваем длину
        
        # Определяем критичность
        critical_keywords = [
            'fail', 'crash', 'timeout', 'disk full', 'out of memory',
            'отказ', 'сбой', 'критический', 'авария'
        ]
        msg_lower = result['message'].lower()
        if result['is_error'] or any(kw in msg_lower for kw in critical_keywords):
            result['is_critical'] = True
        
        # Специальные критичные Event ID для Security
        critical_ids = ['4624', '4625', '4672', '4673', '4688']
        if result['event_id'] in critical_ids:
            result['is_critical'] = True
            result['level'] = 'Security Alert'
        
        return result

def main():
    """Читает строку из stdin, парсит, выводит JSON"""
    parser = WindowsLogParser()
    
    # Читаем всё содержимое из stdin
    raw_input = sys.stdin.read()
    
    if not raw_input.strip():
        print(json.dumps({"error": "No input"}))
        sys.exit(1)
    
    result = parser.parse_event(raw_input)
    print(json.dumps(result, ensure_ascii=False))

if __name__ == "__main__":
    main()